contracts/base/BaseLiquidityManagement.sol

```diff
@@ -96,39 +99,77 @@ contract BaseLiquidityManagement is IBaseLiquidityManagement, SafeCallback {
+
+        // Update position storage (tokensOwed from fees, liquidity added, and new feeGrowthInside).
+        // if callerDelta <= 0, then tokensOwed0 and tokensOwed1 should be zero'd out as all fees were re-invested into a new position.
+        // if callerDelta > 0, then even after re-investing old fees, the caller still has some fees to collect that were not added into the position so they are accounted.
+        uint128 tokensOwed0 = callerDelta0 > 0 ? uint128(callerDelta0) : 0;
+        uint128 tokensOwed1 = callerDelta1 > 0 ? uint128(callerDelta1) : 0;
+
+        position.updateTokensOwed(tokensOwed0, tokensOwed1);
+        position.add(liquidityToAdd);
+        position.updateFeeGrowthInside(feeGrowthInside0X128, feeGrowthInside1X128);
+
+        return toBalanceDelta(callerDelta0, callerDelta1);
+    }
+
+    // Returns the new sanitized delta for the caller.
+    function _accountNewCallerDeltas(
+        BalanceDelta liquidityDelta,
+        BalanceDelta totalFeesAccrued,
+        BalanceDelta callerFeesAccrued,
+        PoolKey memory poolKey
+    ) internal returns (int128 callerDelta0, int128 callerDelta1) {
+        // Only account new caller deltas if there is a difference in between the range's total earned fees and the callers earned fees.
+        (int128 liquidityDelta0, int128 liquidityDelta1) = (liquidityDelta.amount0(), liquidityDelta.amount1());
+        if (totalFeesAccrued == callerFeesAccrued) {
+            return (liquidityDelta0, liquidityDelta1);
         }
 
-        return callerDelta;
+        (int128 totalFeesAccrued0, int128 totalFeesAccrued1) = (totalFeesAccrued.amount0(), totalFeesAccrued.amount1());
+        (int128 callerFeesAccrued0, int128 callerFeesAccrued1) =
+            (callerFeesAccrued.amount0(), callerFeesAccrued.amount1());
+
+        callerDelta0 = totalFeesAccrued0 > callerFeesAccrued0
+            ? _accountNewCallerDelta(liquidityDelta.amount0(), totalFeesAccrued0, callerFeesAccrued0, poolKey.currency0)
+            : liquidityDelta0;
+
+        callerDelta1 = totalFeesAccrued1 > callerFeesAccrued1
+            ? _accountNewCallerDelta(liquidityDelta.amount1(), totalFeesAccrued1, callerFeesAccrued1, poolKey.currency1)
+            : liquidityDelta1;
+    }
+
+    function _accountNewCallerDelta(
+        int128 liquidityDelta,
+        int128 totalFeesAccrued,
+        int128 callerFeesAccrued,
+        Currency currency
+    ) internal returns (int128 callerDelta) {
+        int128 feesAccruedOutsideCaller;
+        unchecked {
+            feesAccruedOutsideCaller = totalFeesAccrued - callerFeesAccrued;
+            callerDelta = liquidityDelta - feesAccruedOutsideCaller;
+        }
+        // We must take the fees accrued to other users on the same range, so the accounting in pool manager is up to date before we take/settle all open deltas.
+        currency.take(manager, address(this), uint128(feesAccruedOutsideCaller), true);
```

Comment:
theres a case where we also want to `take` the remainder fees, but this might break compatability with the other operations?